driver-core/src/main/com/mongodb/client/model/expressions/MqlExpression.java

```diff
@@ -161,6 +170,72 @@ public BooleanExpression lte(final Expression lte) {
         return new MqlExpression<>(ast("$lte", lte));
     }
 
+    public BooleanExpression isBoolean() {
+        return new MqlExpression<>(ast("$type")).eq(of("bool"));
+    }
+
+    @Override
+    public BooleanExpression isBooleanOr(final BooleanExpression or) {
+        return this.isBoolean().cond(this, or);
+    }
+
+    public BooleanExpression isNumber() {
+        return new MqlExpression<>(astWrapped("$isNumber"));
+    }
+
+    @Override
+    public NumberExpression isNumberOr(final NumberExpression or) {
+        return this.isNumber().cond(this, or);
+    }
+
+    public BooleanExpression isString() {
+        return new MqlExpression<>(ast("$type")).eq(of("string"));
+    }
+
+    @Override
+    public StringExpression isStringOr(final StringExpression or) {
+        return this.isString().cond(this, or);
+    }
+
+    public BooleanExpression isDate() {
+        return ofStringArray("date", "timestamp").contains(new MqlExpression<>(ast("$type")));
```

Comment:
If it can be substituted in all the same places, this is fine.

I previously accepted the argument that non-substitutability in eq would be a counterexample, but I am not so sure now. For example, both DateTime dt and plain Date d should count as dates, returning the same values for year and so on. And even though a dt and d might return all the same values (2022, ... ...), they might nevertheless not be equal (because dt is in the 3rd hour), and this is fine.

Reply:
I will remove timestamp (for now, it can be added later)